samples/bluetooth/direct_test_mode/src/transport/hci_uart.c

```diff
@@ -0,0 +1,309 @@
+#include <zephyr/net/buf.h>
+#include <zephyr/bluetooth/hci_types.h>
+#include <zephyr/sys/byteorder.h>
+
+#include "dtm_transport.h"
+#include "hci_uart.h"
+
+#ifndef CONFIG_DTM_TRANSPORT_LOG_LEVEL
+#define CONFIG_DTM_TRANSPORT_LOG_LEVEL CONFIG_LOG_DEFAULT_LEVEL
+#endif
+
+#ifdef CONFIG_DTM_TRANSPORT_HCI
+#define QUEUE_COUNT CONFIG_DTM_HCI_QUEUE_COUNT
+#define QUEUE_SIZE CONFIG_DTM_HCI_QUEUE_SIZE
+#define TX_THREAD_STACK_SIZE CONFIG_DTM_HCI_TX_THREAD_STACK_SIZE
+#define TX_THREAD_PRIORITY CONFIG_DTM_HCI_TX_THREAD_PRIORITY
+#else
+#define QUEUE_COUNT CONFIG_REMOTE_HCI_QUEUE_COUNT
+#define QUEUE_SIZE CONFIG_REMOTE_HCI_QUEUE_SIZE
+#define TX_THREAD_STACK_SIZE CONFIG_REMOTE_HCI_TX_THREAD_STACK_SIZE
+#define TX_THREAD_PRIORITY CONFIG_REMOTE_HCI_TX_THREAD_PRIORITY
+#endif
+
+#define UART_DMA_BUF_SIZE 128
+#define UART_TIMEOUT_US 10000
+
+LOG_MODULE_REGISTER(dtm_hci_uart, CONFIG_DTM_TRANSPORT_LOG_LEVEL);
+
+#define DTM_UART DT_CHOSEN(ncs_dtm_uart)
+static const struct device *hci_uart_dev = DEVICE_DT_GET(DTM_UART);
+
+NET_BUF_POOL_DEFINE(hci_tx_buf, QUEUE_COUNT, QUEUE_SIZE, 0, NULL);
+static K_FIFO_DEFINE(hci_tx_queue);
+
+NET_BUF_POOL_DEFINE(hci_rx_buf, QUEUE_COUNT, QUEUE_SIZE, sizeof(uint8_t), NULL);
+
+enum h4_state {
+	S_TYPE,
+	S_HEADER,
+	S_PAYLOAD
+};
+
+static hci_uart_read_cb dtm_hci_put;
+
+static int hci_hdr_len(uint8_t type)
+{
+	switch (type) {
+	case H4_TYPE_CMD:
+		return sizeof(struct bt_hci_cmd_hdr);
+
+	case H4_TYPE_ACL:
+		return sizeof(struct bt_hci_acl_hdr);
+
+	case H4_TYPE_EVT:
+		return sizeof(struct bt_hci_evt_hdr);
+
+	case H4_TYPE_ISO:
+		return sizeof(struct bt_hci_iso_hdr);
+
+	default:
+		return 0;
+	}
+}
+
+static int hci_pld_len(uint8_t type, uint8_t *hdr)
+{
+	switch (type) {
+	case H4_TYPE_CMD:
+		return ((struct bt_hci_cmd_hdr *)hdr)->param_len;
+
+	case H4_TYPE_ACL:
+		return sys_le16_to_cpu(((struct bt_hci_acl_hdr *)hdr)->len);
+
+	case H4_TYPE_ISO:
+		return sys_le16_to_cpu(((struct bt_hci_iso_hdr *)hdr)->len);
+
+	default:
+		return 0;
+	}
+}
+
+static bool h4_rx_type(uint8_t type)
+{
+	return ((type == H4_TYPE_CMD) | (type == H4_TYPE_ACL) | (type == H4_TYPE_ISO));
+}
+
+static size_t buf_read(struct net_buf *buf, const uint8_t *src, size_t src_len, size_t req_len)
+{
+	size_t len;
+
+	if (req_len > src_len) {
+		len = src_len;
+	} else {
+		len = req_len;
+	}
+
+	if (net_buf_tailroom(buf) < len) {
+		__ASSERT_NO_MSG(false);
+	}
+	net_buf_add_mem(buf, src, len);
+
+	return len;
+}
+
+static void h4_read(const uint8_t *data, size_t offset, size_t len)
+{
+	static enum h4_state state = S_TYPE;
+	static struct net_buf *buf;
+	static uint8_t type;
+	static int rem;
+	size_t read;
+
+	while (len > 0) {
+		switch (state) {
+		case S_TYPE:
+			type = data[offset];
+			offset += sizeof(type);
+			len -= sizeof(type);
+
+			if (h4_rx_type(type)) {
+				rem = hci_hdr_len(type);
+				buf = net_buf_alloc(&hci_rx_buf, K_NO_WAIT);
+				if (!buf) {
+					__ASSERT_NO_MSG(false);
+					/* Out of buffers condition */
```

Comment:
I think it is worth telling the user what went wrong.

Reply:
I checked, the "NO_MSG" assertion prints the line and file name in the assertion, so the user is notified.